net/Socket.cpp

```diff
@@ -1145,63 +1158,160 @@ bool ServerSocket::bind([[maybe_unused]] Type type, [[maybe_unused]] int port)
 #endif
 }
 
+bool ServerSocket::isRetryAcceptError(const int cause)
+{
+    switch(cause)
+    {
+        case EINTR:
+        case EAGAIN:        // == EWOULDBLOCK
+        case ENETDOWN:      // man accept(2): to be treated like EAGAIN
+        case EPROTO:        // man accept(2): to be treated like EAGAIN
+        case ENOPROTOOPT:   // man accept(2): to be treated like EAGAIN
+        case EHOSTDOWN:     // man accept(2): to be treated like EAGAIN
+#ifdef ENONET
+        case ENONET:        // man accept(2): to be treated like EAGAIN (undef on BSD, APPLE, ..)
+#endif
+        case EHOSTUNREACH:  // man accept(2): to be treated like EAGAIN
+        case EOPNOTSUPP:    // man accept(2): to be treated like EAGAIN
+        case ENETUNREACH:   // man accept(2): to be treated like EAGAIN
+        case ECONNABORTED:  // OK
+        case ETIMEDOUT:     // OK, should not happen due ppoll
+            return true;
+        default:
+            return false;
+    }
+}
+
+bool ServerSocket::isRecoverableAcceptError(const int cause)
+{
+    switch(cause)
+    {
+        case EINTR:
+        case EAGAIN:        // == EWOULDBLOCK
+        case ENETDOWN:      // man accept(2): to be treated like EAGAIN
+        case EPROTO:        // man accept(2): to be treated like EAGAIN
+        case ENOPROTOOPT:   // man accept(2): to be treated like EAGAIN
+        case EHOSTDOWN:     // man accept(2): to be treated like EAGAIN
+#ifdef ENONET
+        case ENONET:        // man accept(2): to be treated like EAGAIN (undef on BSD, APPLE, ..)
+#endif
+        case EHOSTUNREACH:  // man accept(2): to be treated like EAGAIN
+        case EOPNOTSUPP:    // man accept(2): to be treated like EAGAIN
+        case ENETUNREACH:   // man accept(2): to be treated like EAGAIN
+        case ECONNABORTED:  // OK
+        case ETIMEDOUT:     // OK, should not happen due ppoll
+        case EMFILE:        // Temporary: No free file descriptors left (per process)
+        case ENFILE:        // Temporary: No free file descriptors left (system)
+        case ENOBUFS:       // Temporary: No free socket memory left (system)
+        case ENOMEM:        // Temporary: No free socket memory left (system)
+            return true;
+        default:
+            return false;
+    }
+}
+
 std::shared_ptr<Socket> ServerSocket::accept()
 {
     // Accept a connection (if any) and set it to non-blocking.
     // There still need the client's address to filter request from POST(call from REST) here.
+    int rc;
 #if !MOBILEAPP
     assert(_type != Socket::Type::Unix);
 
     struct sockaddr_in6 clientInfo;
     socklen_t addrlen = sizeof(clientInfo);
-    const int rc = ::accept4(getFD(), (struct sockaddr *)&clientInfo, &addrlen, SOCK_NONBLOCK | SOCK_CLOEXEC);
+#if TEST_SERVERSOCKET_FAULT_INJECTION
+    const size_t acceptCount = ++TestExternalServerSocketAcceptCount;
+    bool injectedError = false;
+    if (net::Defaults.testExternalServerSocketAcceptSimpleErrorInterval > 0 &&
```

Comment:
please see simulateSocketError for an example of how to do this - could extend the enum paraeter there. We should remove the TEST_SERVERSOCKET... etc.

Reply:
OK